core/src/main/java/org/elasticsearch/index/engine/InternalEngine.java

```diff
@@ -394,30 +396,34 @@ private long checkDeletedAndGCed(VersionValue versionValue) {
     }
 
     @Override
-    public boolean index(Index index) {
-        final boolean created;
+    public void index(Index index) {
         try (ReleasableLock lock = readLock.acquire()) {
             ensureOpen();
             if (index.origin().isRecovery()) {
                 // Don't throttle recovery operations
-                created = innerIndex(index);
+                innerIndex(index);
             } else {
                 try (Releasable r = throttle.acquireThrottle()) {
-                    created = innerIndex(index);
+                    innerIndex(index);
                 }
             }
-        } catch (IllegalStateException | IOException e) {
+        } catch (IllegalStateException | IOException | VersionConflictEngineException e) {
             try {
-                maybeFailEngine("index", e);
+                if (maybeFailEngine("index", e) == false) {
+                    // we set operation specific engine failure that doesn't
```

Comment:
one more issue I have with this - if we go down this route , we need to catch *any* exception , not just the above frew.